### PR TITLE
Revert to token-per-line response file handling

### PR DESCRIPTION
**Pull request by baronfel**
An update to System.CommandLine changed the response file handling support to a different, more cross-platform-consistent logic. However, we cannot take this change because users expect a token-per-line style of processing. As a result, we change the dotnet CLI to use token-per-line handling for response files.

## Description

There are two issues resolved here:
* A change to the defaults in System.CommandLine resulted in response-file parsing breaking across versions. This was resolved by using the S.CL extension points to implement the old, one-token-per-line behavior.
* A user scenario used a version of the MSBuild `--property` option (`/p`) that the .NET SDK didn't have handling for. This was resolved by adding in handling for all unhandled forms of this property.

Fixes https://github.com/dotnet/sdk/issues/26026.

## Customer Impact

Customers that relied on the old response-file behavior typically ran into it in one of two scenarios - the first is related to an internal Brotli compression tool, the workaround here is to publish your application in a directory that contains no spaces. The second involved response files with the `/p` flag - the workaround here is to modify the response file to convert `/p` into one of the recognized forms (`-p`, `--p`, `-property`).

## Regression?

- [X] Yes
- [ ] No

## Risk

- [ ] High
- [X] Medium
- [ ] Low

The change touches response file handling, which can be very broadly used. It does however have testing and reimplements prior behavior.

## Verification

- [X] Manual (required)
- [X] Automated

Manually tested the provided user scenarios, and also added automated testing around the parsing changes.

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [X] N/A

### CODE REVIEW

File: src/Cli/dotnet/Parser.cs [hunk 1]
```diff
@@ -109,6 +109,24 @@ public static Command GetBuiltInCommand(string commandName)
...
         }
 
+        /// <summary>
+        /// Implements token-per-line response file handling for the CLI. We use this instead of the built-in S.CL handling
+        /// to ensure backwards-compatibility with MSBuild.
+        /// </summary>
+        public static bool TokenPerLine(string tokenToReplace, out IReadOnlyList<string> replacementTokens, out string errorMessage) {
+            var filePath = Path.GetFullPath(tokenToReplace);
+            if (File.Exists(filePath)) {
+                replacementTokens = File.ReadAllLines(filePath).Where(line => !line.StartsWith("#")).Select(line => $"\"{line}\"").ToArray();
```
Comment (dsplaisted) [inline src/Cli/dotnet/Parser.cs:119]: How sure are you that this exactly matches the previous behavior?

**Review by dsplaisted — APPROVED** [review]

File: src/Cli/dotnet/Parser.cs (hunk 1 above)
Comment (baronfel) [reply src/Cli/dotnet/Parser.cs:119]: working through some of the details of that with @jonsequitur right now, will report back

File: src/Tests/dotnet.Tests/ParserTests/ResponseFileTests.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.DotNet.Tests.ParserTests
+{
+    public class ResponseFileTests
+    {
+        [Fact]
+        public void Can_safely_expand_response_file_lines() {
+            var tempFilePath = Path.GetTempFileName();
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tests/dotnet.Tests/ParserTests/ResponseFileTests.cs:19]: thanks, will do

File: src/Cli/dotnet/OptionForwardingExtensions.cs [hunk 3]
```diff
@@ -19,7 +19,11 @@ public static class OptionForwardingExtensions
         public static ForwardedOption<T> ForwardAsSingle<T>(this ForwardedOption<T> option, Func<T, string> format) => option.SetForwardingFunction(format);
 
         public static ForwardedOption<string[]> ForwardAsProperty(this ForwardedOption<string[]> option) => option
-            .SetForwardingFunction((optionVals) => optionVals.SelectMany(optionVal => new string[] { $"{option.Aliases.FirstOrDefault()}:{optionVal.Replace("roperty:", string.Empty)}" }));
+            .SetForwardingFunction((optionVals) =>
+                optionVals
+                    .Select(optionVal => optionVal.Replace(";", "%3B")) // must escape semicolon-delimited property values when forwarding them to MSBuild
```
Comment (baronfel) [inline src/Cli/dotnet/OptionForwardingExtensions.cs:24]: this is critical - without escaping (which we do in several other places in the codebase we probably should try to unify), msbuild will eat the `;` character. This is a longstanding issue that was helpfully pointed out by @janv8000 [here](https://github.com/dotnet/sdk/issues/26026#issuecomment-1161431464) that I boorishly disregarded. many thanks, friend!

as a result now, we consistently handle all msbuild-known property option forms.

File: src/Cli/dotnet/Parser.cs [hunk 4]
```diff
@@ -109,6 +109,39 @@ public static Command GetBuiltInCommand(string commandName)
                 .FirstOrDefault(c => c.Name.Equals(commandName, StringComparison.OrdinalIgnoreCase));
         }
 
+        /// <summary>
+        /// Implements token-per-line response file handling for the CLI. We use this instead of the built-in S.CL handling
+        /// to ensure backwards-compatibility with MSBuild.
+        /// </summary>
+        public static bool TokenPerLine(string tokenToReplace, out IReadOnlyList<string> replacementTokens, out string errorMessage) {
```
Comment (baronfel) [inline src/Cli/dotnet/Parser.cs:116]: this is the core change that gets us back to parity with the response file parsing from the older S.CL version.  We get a presumed file path, and if that file path isn't an actual path error out to the user (S.CL did this previously). Then, we read the lines, strip out data from any # tokens onward, and drop all subsequently-empty lines.

File: src/Cli/dotnet/commands/dotnet-msbuild/MSBuildForwardingApp.cs [hunk 5]
```diff
@@ -57,6 +57,8 @@ public MSBuildForwardingApp(IEnumerable<string> argsToForward, string msbuildPat
             }
         }
 
+        public IEnumerable<string> MSBuildArguments { get { return _forwardingAppWithoutLogging.GetAllArguments(); } }
```
Comment (baronfel) [inline src/Cli/dotnet/commands/dotnet-msbuild/MSBuildForwardingApp.cs:60]: test helper that helped me see the property-forwarding problem.

File: src/Tests/dotnet.Tests/ParserTests/MSBuildArgumentCommandLineParserTests.cs [hunk 6]
```diff
@@ -31,9 +31,10 @@ public void MSBuildArgumentsAreForwardedCorrectly(string[] arguments, bool build
             RestoringCommand command = buildCommand ? 
                 BuildCommand.FromArgs(arguments) : 
                 PublishCommand.FromArgs(arguments);
+            var expectedArguments = arguments.Select(a => a.Replace("-p:", "-property:"));
```
Comment (baronfel) [inline src/Tests/dotnet.Tests/ParserTests/MSBuildArgumentCommandLineParserTests.cs:34]: we expand the `-p` form to `-property` now for consistency, so the expectations needed updating

File: src/Tests/dotnet.Tests/ParserTests/ResponseFileTests.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                "build",
+                "a b",
+                "-p",
+                "VSTestTestAdapterPath=\".;/opt/buildagent/plugins/dotnet/tools/vstest15\""
+            };
+
+            Log.WriteLine($"MSbuild Args are {string.Join(" ", bc.MSBuildArguments)}");
+            Log.WriteLine($"Parse Diagram is {parseResult.Diagram()}");
+            Log.WriteLine($"Token string is {tokenString}");
+            tokens.Skip(1).Should().BeEquivalentTo(tokenized);
```
Comment (baronfel) [inline src/Tests/dotnet.Tests/ParserTests/ResponseFileTests.cs:49]: note - probably need a test with # and empty strings still, but this validates the user-provided use case tokenizes correctly.

File: src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs [hunk 8]
```diff
@@ -56,7 +56,7 @@ public void MsbuildInvocationIsCorrect(string[] args, string expectedAdditionalA
 
         [Theory]
         [InlineData(new string[] { "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
-        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-p:TargetFramework=tfm")]
+        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
```
Comment (baronfel) [inline src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs:59]: since all `-p` forms are normalized to `-property`, this needed updating as well.

**Review by baronfel — COMMENTED** [review]
A few notes on the updates from today/consulting with @jonsequitur.

File: src/Cli/dotnet/Parser.cs (hunk 1 above)
Comment (baronfel) [reply src/Cli/dotnet/Parser.cs]: wrapping in quotes is no longer necessary - this was an artifacts of some msbuild ;-gobbling that I've now addressed.

File: src/Cli/dotnet/CommonOptions.cs [hunk 9]
```diff
@@ -17,7 +17,7 @@ namespace Microsoft.DotNet.Cli
     internal static class CommonOptions
     {
         public static Option<string[]> PropertiesOption =
-            new ForwardedOption<string[]>(new string[] { "-property", "/p" })
+            new ForwardedOption<string[]>(new string[] { "-property", "/p", "-p" })
```
[the remark this answers is not in the document]
Comment (dsplaisted) [reply src/Cli/dotnet/CommonOptions.cs:20]: What about `/property`?  I think that would work too.

File: src/Cli/dotnet/OptionForwardingExtensions.cs [hunk 10]
```diff
@@ -19,7 +19,11 @@ public static class OptionForwardingExtensions
...
+                    .SelectMany(optionVal => new string[] { $"{option.Aliases.FirstOrDefault()}:{optionVal.Replace("roperty:", string.Empty)}" })
```
Comment (dsplaisted) [inline src/Cli/dotnet/OptionForwardingExtensions.cs:25]: Will this work correctly if you do something like `-p:MyCrazyProperty=property:value`, or will it remove the `roperty:` portion of the property value?

**Review by dsplaisted — APPROVED** [review]

File: src/Cli/dotnet/OptionForwardingExtensions.cs (hunk 10 above)
Comment (baronfel) [reply src/Cli/dotnet/OptionForwardingExtensions.cs:25]: good catch - I'll add some tests around this forwarding function specifically (by extracting it to a named function and slamming it with different inputs), but IMO we should be canonicalizing all of the different input shapes into `-property:<user input>` for MSBuild (meaning that we use the most verbose form available).

**Comment by baronfel** [comment]
I could add `/property` as well, I thought about that last night - ideally we'd accept all forms of the property option that MSbuild accepts, _as well_ as the more posix-standard `--property` long form and `-p` short form.

File: src/Tests/dotnet.Tests/ParserTests/ResponseFileTests.cs (hunk 7 above)
Comment (baronfel) [reply src/Tests/dotnet.Tests/ParserTests/ResponseFileTests.cs:49]: added this test now.

**Comment by baronfel** [comment]
@dsplaisted if you wanna take another look that'd be cool :) I added full support for the different permutations of `property` that MSBuild supports (`--`, `-`, `/`), and fleshed out testing for the token replacement for response files as well as the MSBuild property forwarding since the last time you looked.

**Review by dsplaisted — APPROVED** [review]

**Comment by BradBarnich** [comment]
will this change fix to allow handling a comma in a quoted string in an RSP file? I didn't see this case covered in the tests

such as:
```
/p:Copyright="Copyright (C) CareEvolution, Inc. 2022"
```
another example in this comment: https://github.com/dotnet/sdk/issues/26026#issuecomment-1157522720

**Comment by baronfel** [comment]
@BradBarnich thanks for checking! the response file part is parsed correctly, but the value gets mutilated during the forwarding to MSBuild. I will add additional test cases and fixes for this now.

**Comment by baronfel** [comment]
Actually this _did_ fix parameters with spaces and quotes, and I was just comparing to the wrong collection in the tests. Updated and we should be good to go!
